### PR TITLE
Use RFC3339 for Timestamp

### PR DESCRIPTION
**Pull request by leonardehrenfried**
#### **If you are new to the specification, please introduce yourself (name and organization/link to GBFS). It’s helpful to know who we're chatting with!**  

I'm a core developer of OpenTripPlanner, a multimodal transit routing engine with GBFS support.

#### **What problem does your proposal solve? Please begin with the relevant issue number. If there is no existing issue, please also describe alternative solutions you have considered.**

Right now GBFS uses seconds since the epoch as a timestamp, for example in the last_updated field. While this practice is pretty wide-spread, it makes debugging feeds a little harder than necessary.

Every time I need to check the last_updated field I need to copy the epoch seconds, google "epoch to time converter", paste the seconds to get a usable date time.

While this isn't an impossible task, I think there are alternatives that are both human-readable and unique so that machines can identify a universal point in time.

#### **What is the proposal?**

Replacing POSIX time with RFC3339 makes a much more pleasant debugging experience for humans and doesn't add ambiguity for computers.

An example of RFC3339 time is `2023-07-17T13:45:29+02:00`.

#### **Is this a breaking change?**
- [X] Yes 
- [ ] No
- [ ] Unsure

#### **Which files are affected by this change?**

All of them. :flushed: 

#### Timezone

~It was suggested in #519 that we should limit the offset/timezone to UTC, so an offset of 0 or a `Z` at the end.~

~Personally, I'm not in favour of this restriction as you should use a proper date parsing library which handles all offsets, but as time is tight I would like to keep the discussion to a minimum and not let this point block the proposal.~

EDIT: Consensus was reached to include the offset/timezone in this PR.

### CODE REVIEW

**Comment by josee-sabourin** [comment]
Tagging those who were active in #519 
@mplsmitch @sven4all @benwedge @testower @futuretap

**Comment by benwedge** [comment]
> Timezone
It was suggested in https://github.com/MobilityData/gbfs/issues/519 that we should limit the offset/timezone to UTC, so an offset of 0 or a Z at the end.

>Personally, I'm not in favour of this restriction as you should use a proper date parsing library which handles all offsets, but as time is tight I would like to keep the discussion to a minimum and not let this point block the proposal.

IMO we should just get it right the first time. If it's going to be a breaking change and every language has a library that handles this nicely why not just do it?

**Comment by leonardehrenfried** [comment]
> IMO we should just get it right the first time. If it's going to be a breaking change and every language has a library that handles this nicely why not just do it?

That would of course be okay for me.

**Comment by testower** [comment]
I agree with @benwedge let us not do it half way.

**Comment by futuretap** [comment]
I support the proposal, including variable offsets.

**Comment by leonardehrenfried** [comment]
Thanks all for chiming in. I will wait until the end of today and if no one shows up voicing opposition to variable offsets I will update the PR.

**Comment by leonardehrenfried** [comment]
Since no-one has shown up who is against variable offsets, I have updated the PR.

File: gbfs.md [hunk 1]
```diff
@@ -1275,8 +1275,8 @@ Field Name | REQUIRED | Type | Defines
         ],
         "times": [
           {
-            "start": 1604448000,
-            "end": 1604674800
+            "start": "2023-07-17T13:34:13+02:00",
+            "end": "2023-07-18T13:34:13Z"
```
Comment (mplsmitch) [inline gbfs.md]: local offset  for "end" should be +2:00

File: gbfs.md [hunk 2]
```diff
@@ -1426,8 +1426,8 @@ See examples below.
                 "language": "en"
               }
             ],
-            "start": 1593878400,
-            "end": 1593907260,
+            "start": "2023-07-17T13:34:13+02:00",
+            "end": "2024-07-17T13:34:13Z",
```
Comment (mplsmitch) [inline gbfs.md]: local offset  for "end" should be +2:00

File: gbfs.md (hunk 1 above)
Comment (leonardehrenfried) [reply gbfs.md]: You're right, that's a search & replace missing this case. Corrected in 348929606c7541cbcfac90490700c6178ddcd4d0.

File: gbfs.md (hunk 2 above)
Comment (leonardehrenfried) [reply gbfs.md]: You're right, that's a search & replace missing this case. Corrected in 348929606c7541cbcfac90490700c6178ddcd4d0.

**Review by mplsmitch — APPROVED** [review]

**Comment by josee-sabourin** [comment]
I (as proxy for @leonardehrenfried who is on vacation) hereby call a vote on this proposal. Voting will be open for 10 full calendar days until 11:59PM UTC on August 5th.
Please vote for or against the proposal, and include the organization for which you are voting in your comment.
Please note if you can commit to implementing the proposal.

**Comment by testower** [comment]
Entur votes in favor of this proposal and will implement it

**Comment by futuretap** [comment]
We (Where To?) are in favor, too, and will implement it.

**Comment by AntoineAugusti** [comment]
transport.data.gouv.fr is in favour and will implement it (for feeds we consume)

**Comment by benwedge** [comment]
Joyride (provider) is in favour and will implement this when we move to 3.x

**Comment by leonardehrenfried** [comment]
Not sure if my vote counts but OpenTripPlanner will implement this via Entur's Java library.

**Comment by josee-sabourin** [comment]
This vote has now closed, and it passes!

Votes in favor:
Entur (consumer)
Whereto.app (consumer)
Transport.data.gouv (consumer)
Joyride (producer)
 
There were no votes against.

**Comment by richfab** [comment]
Datetime is currently `ISO 8601` notation in the spec. Should Datetime be replaced by Timestamp (`RFC3339`)? 

Datetime definition in the spec:
https://github.com/MobilityData/gbfs/blob/master/gbfs.md?plain=1#L195

Datetime only usage is in vehicle_status.json for `available_until` field:
https://github.com/MobilityData/gbfs/blob/master/gbfs.md?plain=1#L956

**Comment by testower** [comment]
Good catch @richfab - I think for most practical purposes they are the same, but the spec should definitely be consistent. The ISO is very broad, so I would be in favor of restricting it to the RFC (which is a profile of the ISO standard?)

**Comment by leonardehrenfried** [comment]
My suggestion is to combine the two, so all of them will be of type Timestamp or DateTime.

I have a slight personal preference for the name DateTime but I'm fine with both.

**Comment by leonardehrenfried** [comment]
@josee-sabourin What is the process of merging this PR?

I would prefer to merge this and then make a follow up where I combine the two types since that would be a non-breaking change.

**Comment by josee-sabourin** [comment]
@leonardehrenfried currently we are waiting to merge pending a few more potential breaking changes to limit the number of v3.0 release candidates we have.

**Comment by hbruch** [comment]
Note that [RFC3339](https://www.rfc-editor.org/rfc/rfc3339) allows `T` and `Z` also to be written in lowercase, `T` might also be left out. Should the spec prescribe uppercase `T` and `Z` or allow these variations? In the latter case, the regex in gbfs-json-schema of course should reflect this.

**Comment by richfab** [comment]
Hello @hbruch,

Great question. For reference, this is the extract from the [RFC3339](https://www.rfc-editor.org/rfc/rfc3339) spec:

>NOTE: Per [[ABNF](https://www.rfc-editor.org/rfc/rfc3339#ref-ABNF)] and ISO8601, the "T" and "Z" characters in this syntax may alternatively be lower case "t" or "z" respectively.
This date/time format may be used in some environments or contexts that distinguish between the upper- and lower-case letters 'A'-'Z' and 'a'-'z' (e.g. XML).  Specifications that use this format in such environments MAY further limit the date/time syntax so that the letters 'T' and 'Z' used in the date/time syntax must always be upper case.  Applications that generate this format SHOULD use upper case letters.
NOTE: ISO 8601 defines date and time separated by "T". Applications using this syntax may choose, for the sake of readability, to specify a full-date and full-time separated by (say) a space character.

What is everyone's opinion about:
1. Restricting `T` and `Z` to upper case or not?
2. Allowing `T` to be left out or not?

Thank you!

**Comment by leonardehrenfried** [comment]
I would be fine with not supporting the full flexibility of the spec and require them both to be mandatory and upper case. (Z can of course be replaced with an offset).

**Comment by futuretap** [comment]
We're strongly preferring to restrict `T` and `Z` to upper case and require it not to be left out. Technical reason: Swift `Decodable` requires this strict format.

**Review by richfab — APPROVED** [review]
LGTM

File: gbfs.md [hunk 3]
```diff
@@ -221,7 +221,7 @@ Example: The `rental_methods` field contains values `creditcard`, `paypass`, etc
 * Phone Number *as of v3.0-RC* - Phone number in [E.164](https://www.itu.int/rec/T-REC-E.164-201011-I/en) format. The phone number MUST start with a "+". The characters following the "+" MUST be integers and MUST NOT contain any hyphens, spaces or parentheses.
 * String - Can only contain text. Strings MUST NOT contain any formatting codes (including HTML) other than newlines.
 * Time - Service time in the HH:MM:SS format for the time zone indicated in `system_information.json` (00:00:00 - 47:59:59). Time can stretch up to one additional day in the future to accommodate situations where, for example, a system was open from 11:30pm - 11pm the next day (23:30:00-47:00:00).
-* Timestamp - Timestamp fields MUST be represented as integers in POSIX time (representing the number of seconds since January 1st 1970 00:00:00 UTC).
+* Timestamp - Timestamp fields MUST be represented as strings in [RFC3339 format](https://www.rfc-editor.org/rfc/rfc3339), for example `2023-07-17T13:34:13+02:00`.
```
Comment (richfab) [inline gbfs.md]: As of v3.0-RC2

```suggestion
* Timestamp - Timestamp fields MUST be represented as strings in [RFC3339 format](https://www.rfc-editor.org/rfc/rfc3339), for example `2023-07-17T13:34:13+02:00`. *(as of v3.0-RC2)*
```

**Pull request merged**